src/compute-client/src/controller/introspection.rs

```diff
@@ -0,0 +1,370 @@
+    fn log_typ(log: LogVariant) -> RelationType {
+        log.desc().typ().clone()
+    }
+
+    /// Imports a logging index into the builder state, returns the corresponding `Get` plan.
+    fn import_log<T>(&mut self, log: LogVariant, mfp: MapFilterProject) -> Plan<T> {
+        self.used_indexes.insert(log);
+
+        let index_id = self.index_ids[&log];
+        let key = Self::log_key(log);
+        let column_types = Self::log_typ(log).column_types;
+        let arity = column_types.len();
+        let (permutation, thinning) = permutation_for_arrangement(&key, arity);
+
+        Plan::Get {
+            id: Id::Global(index_id),
+            keys: AvailableCollections::new_arranged(
+                vec![(key.clone(), permutation, thinning)],
+                Some(column_types),
+            ),
+            plan: GetPlan::Arrangement(key, None, mfp),
+            lir_id: self.allocate_lir_id(),
+        }
+    }
+
+    /// Transforms the given `plan` into a `DataflowDescription` for an introspection subscribe.
+    fn dataflow_description<T: Timestamp>(
+        &self,
+        plan: Plan<T>,
+        desc: RelationDesc,
+    ) -> DataflowDescription<Plan<T>, (), T> {
+        let mut index_imports = BTreeMap::new();
+        for log in &self.used_indexes {
+            let index_id = self.index_ids[log];
+            let import = IndexImport {
+                desc: IndexDesc {
+                    on_id: index_id,
+                    key: Self::log_key(*log),
+                },
+                typ: Self::log_typ(*log),
+                monotonic: false,
+            };
+            index_imports.insert(index_id, import);
+        }
+
+        let sink_desc = ComputeSinkDesc {
+            from: self.view_id,
+            from_desc: desc,
+            connection: ComputeSinkConnection::Subscribe(SubscribeSinkConnection {}),
+            with_snapshot: true,
+            up_to: Default::default(),
+            non_null_assertions: Default::default(),
+            refresh_schedule: None,
+        };
+
+        DataflowDescription {
+            source_imports: Default::default(),
+            index_imports,
+            objects_to_build: vec![BuildDesc {
+                id: self.view_id,
+                plan,
+            }],
+            index_exports: Default::default(),
+            sink_exports: [(self.sink_id, sink_desc)].into(),
+            as_of: None,
+            until: Default::default(),
+            initial_storage_as_of: None,
+            refresh_schedule: None,
+            debug_name: format!("introspection-subscribe-{}", self.sink_id),
+        }
+    }
+}
+
+/// Builds a dataflow for the `ComputeErrorCounts` subscribe.
+///
+/// The return dataflow corresponds to this SQL query:
+///
+/// ```sql
+/// SELECT export_id, sum(count)
+/// FROM <ComputeLog::ErrorCount>
+/// GROUP BY export_id
+/// ```
+fn build_subscribe_error_counts<T: Timestamp>(
```

Comment:
I understand the complaint about writing dataflows as LIR constructs. They are hard to read, especially compared to the much simpler SQL queries. I'm not sure I understand this part though:

> It locks in assumptions about how we'll use the system, and any change requires an updated version of Materialize.

This is true for almost any change we make to the database, and given that our release cycle is short I don't consider it much of a problem either. Why would we need to be able to update unified introspection sources at a different cadence than the per-replica ones?

> Also, we don't get the optimizer to improve our queries.

This can be a drawback or an advantage, depending on how much we expect the optimizer to produce better LIR than what we can write by hand.

Reply:
Alternative implementation that moves the management of introspection subscribes into the coordinator: https://github.com/MaterializeInc/materialize/pull/27795